ql/ql/src/codeql_ql/StructuredLogs.qll

```diff
@@ -317,12 +321,137 @@ module KindPredicatesLog {
     Predicate getPredicate() { result = getPredicateFromPosition(this.getPosition()) }
   }
 
+  /** Gets the `index`'th event that's evaluated by `recursive`. */
+  private InLayer layerEventRank(ComputeRecursive recursive, int index) {
+    result =
+      rank[index + 1](InLayer cand, int startline, string filepath |
+        cand.getComputeRecursiveEvent() = recursive and
+        cand.hasLocationInfo(filepath, startline, _, _, _)
+      |
+        cand order by filepath, startline
+      )
+  }
+
+  /**
+   * Gets the first predicate that's evaluated in an iteration
+   * of the SCC computation rooted at `recursive`.
+   */
+  private InLayer firstPredicate(ComputeRecursive recursive) {
+    result = layerEventRank(recursive, 0)
+  }
+
+  /**
+   * Gets the last predicate that's evaluated in an iteration
+   * of the SCC computation rooted at `recursive`.
+   */
+  private InLayer lastPredicate(ComputeRecursive recursive) {
+    exists(int n |
+      result = layerEventRank(recursive, n) and
+      not exists(layerEventRank(recursive, n + 1))
```

Comment:
Hmm... Would it be possible/better to express this using `max`? There's something tickling at the back of my mind that I can't quite articulate.

Reply:
We could do it with a max as well, yeah. In my experience this join + antijoin pattern patterns better than an aggregation, though.

Reply:
But to be fair I didn't actually test it in this case 😂.

Reply:
Interesting. I'll trust your experience, then. 🙂 

(My concern was based on [this](https://github.com/github/codeql/commit/22aa4c93799af788eb27feeae6c690926fafaa3c) bad join in the Python regex libraries, though I guess that's a somewhat different situation.)

Reply:
Ah, I see. Yeah, I guess the bad join which that commit is because the `<` and `>` filtering appear too late in the pipeline (which hopefully shouldn't be relevant here).